File: Proxy/proxy.js
class apiSystem{
    get(){
        throw new Error("api call not implemented");
    }
}

class realAPISystem extends apiSystem{
    get(endpoint){
        return "api call result";
    }
}

class proxyAPISystem extends apiSystem{
    constructor(userId){
        super();
        this.apiObj=new realAPISystem();
        this.user = userId;
        this.cache = new Map();
    }

    get(endpoint){
        if(this.user !== 'admin'){
            return "user not allowed to make call";
        }

        console.log("req -> ",endpoint);

        if(this.cache.has(endpoint)){
            return `map res-> ${this.cache.get(endpoint)}`;
        }

        const res =this.apiObj.get(endpoint);
        
        this.cache.set(endpoint,res);
        return res;
    }
}

const client = new proxyAPISystem('user');
console.log(client.get('/user'));

const client2 = new proxyAPISystem('admin');
console.log(client2.get('/user'));
console.log(client2.get('/user'));

File: Proxy/Proxy.java
package Proxy;

import java.util.HashMap;

class ApiSystem {
    public String get(String endpoint){
        return "implement get()";
    }    
}
class RealApiSystem extends ApiSystem {
    public String get(String endpoint){
        return "api call";
    }
}
class proxyApiSystem extends ApiSystem{
    String user;
    HashMap<String,String> cache;
    ApiSystem obj;
    proxyApiSystem(String user){
        this.obj = new RealApiSystem();
        this.cache = new HashMap<>();
        this.user = user;
    }
    public String get(String endpoint){
        if(!this.user.equals("admin")){
            return "user not allowed to make call";
        }

        System.out.println("req -> "+endpoint);

        if(this.cache.containsKey(endpoint)){
            return cache.get(endpoint);
        }

        String res =this.obj.get(endpoint);
        
        this.cache.put(endpoint,res);
        return res;        
    }
}
public class Proxy {
    public static void main(String[] args) {
        proxyApiSystem client = new proxyApiSystem("user");
        System.out.println(client.get("/call"));
        proxyApiSystem client2 = new proxyApiSystem("admin");
        System.out.println(client2.get("/call"));
    }
}
